File: AesGcm256Pbkdf2StringEncryption/AesGcm256Pbkdf2StringEncryption.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

public class Program {
	public static void Main() {
		Console.WriteLine("AES GCM 256 String encryption with PBKDF2 derived key");

		string plaintext = "The quick brown fox jumps over the lazy dog";
		Console.WriteLine("plaintext: " + plaintext);
		string password = "secret password";

		// encryption
		Console.WriteLine("\n* * * Encryption * * *");
		String ciphertextBase64 = aesGcmPbkdf2EncryptToBase64(password, plaintext);
		Console.WriteLine("ciphertext (Base64): " + ciphertextBase64);
		Console.WriteLine("output is (Base64) salt : (Base64) nonce : (Base64) ciphertext : (Base64) gcmTag");

		// decryption
		Console.WriteLine("\n* * * Decryption * * *");
		String ciphertextDecryptionBase64 = ciphertextBase64;
		Console.WriteLine("ciphertext (Base64): " + ciphertextDecryptionBase64);
		Console.WriteLine("input is (Base64) salt : (Base64) nonce : (Base64) ciphertext : (Base64) gcmTag");
		String decryptedtext = aesGcmPbkdf2DecryptFromBase64(password, ciphertextDecryptionBase64);
		Console.WriteLine("plaintext: " + decryptedtext);
	}

	static string aesGcmPbkdf2EncryptToBase64(string password, string data) {
		int PBKDF2_ITERATIONS = 15000;
		byte[] salt = GenerateSalt32Byte();
		byte[] key = new byte[32];
		using(var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2_ITERATIONS, HashAlgorithmName.SHA256)) {
			key = deriveBytes.GetBytes(32);
		}
		byte[] plaintext = Encoding.UTF8.GetBytes(data);
		byte[] gcmTag = new byte[16];
		byte[] nonce = GenerateRandomNonce();
		byte[] cipherText = new byte[plaintext.Length];
		byte[] associatedData = new byte[0];
		using(var cipher = new AesGcm(key)) {
			cipher.Encrypt(nonce, plaintext, cipherText, gcmTag, associatedData);
			return Base64Encoding(salt) + ":" + Base64Encoding(nonce) + ":" + Base64Encoding(cipherText) + ":" + Base64Encoding(gcmTag);
		}
	}

	static string aesGcmPbkdf2DecryptFromBase64(string password, string data) {
		string decryptedtext;
		int PBKDF2_ITERATIONS = 15000;
		String[] parts = data.Split(':');
		byte[] salt = Base64Decoding(parts[0]);
		byte[] key = new byte[32];
		using(var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2_ITERATIONS, HashAlgorithmName.SHA256)) {
			key = deriveBytes.GetBytes(32);
		}
		byte[] associatedData = new byte[0];
		byte[] ciphertext = Base64Decoding(parts[2]);
		byte[] decryptedData = new byte[ciphertext.Length];
		using(var cipher = new AesGcm(key)) {
			cipher.Decrypt(Base64Decoding(parts[1]), ciphertext, Base64Decoding(parts[3]), decryptedData, associatedData);
			decryptedtext = Encoding.UTF8.GetString(decryptedData, 0, decryptedData.Length);
			return decryptedtext;
		}
	}

	static byte[] GenerateSalt32Byte() {
		RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
		byte[] salt = new byte[32];
		rngCsp.GetBytes(salt);
		return salt;
	}

	static byte[] GenerateRandomNonce() {
		RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
		byte[] nonce = new byte[12];
		rngCsp.GetBytes(nonce);
		return nonce;
	}

	static string Base64Encoding(byte[] input) {
		return Convert.ToBase64String(input);
	}

	static byte[] Base64Decoding(String input) {
		return Convert.FromBase64String(input);
	}
}

File: AesGcm256Pbkdf2StringEncryption/AesGcm256Pbkdf2StringEncryption.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
  "encoding/base64"
	"fmt"
  "io"
	"strings"
	"golang.org/x/crypto/pbkdf2"
)

func main() {
  fmt.Printf("AES GCM 256 String encryption with PBKDF2 derived key\n")

  plaintext := string("The quick brown fox jumps over the lazy dog")
  fmt.Printf("plaintext: " + plaintext + "\n")
  password := string("secret password")

  fmt.Printf("\n* * * Encryption * * *\n");
  ciphertextBase64 := string(AesGcmPbkdf2EncryptToBase64(password, plaintext))
  fmt.Printf("ciphertext: " + ciphertextBase64 + "\n");
  fmt.Printf("output is (Base64) salt : (Base64) nonce : (Base64) ciphertext : (Base64) gcmTag\n");
	
  fmt.Printf("\n* * * Decryption * * *\n");
  ciphertextDecryptionBase64 := string(ciphertextBase64);
  fmt.Printf("ciphertext: " + ciphertextDecryptionBase64 + "\n");
  fmt.Printf("input is (Base64) salt : (Base64) nonce : (Base64) ciphertext : (Base64) gcmTag\n");
  decryptedtext := string(AesGcmPbkdf2DecryptFromBase64(password, ciphertextDecryptionBase64))
  fmt.Printf("decryptedtext: " + decryptedtext)
}

func AesGcmPbkdf2EncryptToBase64(passphrase string, data string)(string) {
  PBKDF2_ITERATIONS := int(15000)
  salt := []byte(GenerateSalt32Byte())
  // derive key
  key := []byte(pbkdf2.Key([]byte(passphrase), salt, PBKDF2_ITERATIONS, 32, sha256.New))
  // encrypt
  plaintext := []byte(data)
  nonce := []byte(GenerateRandomNonce())
	block, err := aes.NewCipher(key)
	if err != nil {
		panic(err.Error())
	}
	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		panic(err.Error())
	}
  ciphertext := aesGCM.Seal(nil, nonce, plaintext, nil)
  // ciphertext hold the ciphertext | gcmTag
  ciphertextWithTagLength := int(len(ciphertext))
  ciphertextWithoutTag, gcmTag := ciphertext[:(ciphertextWithTagLength-16)], ciphertext[(ciphertextWithTagLength-16):]
  ciphertextWithoutTagBase64 := string(Base64Encoding(ciphertextWithoutTag))
  saltbase64 := string(Base64Encoding(salt))
  nonceBase64 := string(Base64Encoding(nonce))
  gcmTagBase64 := string(Base64Encoding(gcmTag))
  ciphertextCompleteBase64 := string(saltbase64 + ":" + nonceBase64 + ":" + ciphertextWithoutTagBase64 + ":" + gcmTagBase64) 
  return ciphertextCompleteBase64
}

func AesGcmPbkdf2DecryptFromBase64(passphrase string, ciphertextCompleteBase64 string)(string) {
  PBKDF2_ITERATIONS := int(15000)
  data := strings.Split(ciphertextCompleteBase64, ":") 
  salt := []byte(Base64Decoding(data[0]))
  nonce := []byte(Base64Decoding(data[1]))
  ciphertext := []byte(Base64Decoding(data[2]))
  gcmTag := []byte(Base64Decoding(data[3]))
  ciphertextCompleteLength := int(len(ciphertext) + len(gcmTag))
  ciphertextComplete := make([]byte, ciphertextCompleteLength)
  ciphertextComplete = append(ciphertext[:], gcmTag[:]...)
  // derive key
  key := []byte(pbkdf2.Key([]byte(passphrase), salt, PBKDF2_ITERATIONS, 32, sha256.New))
  // decrypt
  block, err := aes.NewCipher(key)
	if err != nil {
		panic(err.Error())
	}
	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		panic(err.Error())
	}
  plaintext, err := aesGCM.Open(nil, nonce, ciphertextComplete, nil)
	if err != nil {
		panic(err.Error())
	}
  return string(plaintext)
}

func GenerateSalt32Byte()([]byte) {
  salt := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
	  panic(err.Error())
	}
  return salt
}

func GenerateRandomNonce()([]byte) {
  nonce := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
	  panic(err.Error())
	}
  return nonce
}

func Base64Encoding(input []byte)(string) {
  return base64.StdEncoding.EncodeToString(input)
}

func Base64Decoding(input string)([]byte) {
  data, err := base64.StdEncoding.DecodeString(input)
	if err != nil {
		return data
	}
  return data
}
